load unit testing post (#1890)
dbt-labs/docs.getdbt.com: website/blog/2022-08-22-unit-testing-dbt-package.md

---
title: "An introduction to unit testing your dbt Packages"
description: "Traditionally, integration tests have been the primary strategy for testing dbt Packages. In this post, Yu Ishikawa walks us through adding in unit testing as well."
slug: unit-testing-dbt-packages
authors: [yu_ishikawa]
tags: [dbt tutorials]
hide_table_of_contents: false

date: 2022-08-25
is_featured: true
---

It’s important to be able to test any dbt Project, but it’s even more important to make sure you have robust testing if you are developing a [dbt Package](https://docs.getdbt.com/docs/building-a-dbt-project/package-management).

> I love dbt packages, because it makes it easy to extend dbt’s functionality and create reusable analytics resources. Even better, we can find and share dbt packages which others developed, finding great packages in [dbt hub](https://hub.getdbt.com/). However, it is a bit difficult to develop complicated dbt macros, because dbt on top of [Jinja2](https://palletsprojects.com/p/jinja/) is lacking some of the functionality you’d expect for software development - like unit testing.

Reviewer: ```suggestion
I love dbt packages, because they make it easy to extend dbt’s functionality and create reusable analytics resources. Even better, we can find and share dbt packages which others have developed, finding great packages in the [dbt Package Hub](https://hub.getdbt.com/). However, it is a bit difficult to develop complicated dbt macros, because dbt on top of [Jinja2](https://palletsprojects.com/p/jinja/) is lacking some of the functionality you’d expect for software development—like unit testing.
```

To use the integration tests - you’d simply run `dbt test` within the `integration_tests` directory. The tests execute as normal - meaning you can use your favorite methods of running CI against your dbt project to ensure that your integration tests are passing.

Integration tests can help give you peace of mind that your package is performing as expected - but they have some drawbacks. Macros and generic tests frequently call other macros and the deeper dependency calls get, the more difficult it becomes to debug your macros using only integration tests.

In this scenario it can be helpful to go beyond integration tests and implement unit tests for your macros. These unit tests can be run with a [dbt run operation](https://docs.getdbt.com/reference/commands/run-operation). Let’s take a look at a quick example of how this can be done.

> Consider a dbt package called dbt_sample_package . We would like to implement a simple macro to create a string literal from a string text in a macro named `to_literal` in the file `macros/to_literal.sql`.

Reviewer: ```suggestion
Consider a dbt package called `dbt_sample_package` . We would like to implement a simple macro to create a string literal from a string text in a macro named `to_literal` in the file `macros/to_literal.sql`.
```

```sql
- macros/to_literal.sql
{% to_literal(text) %}
'{{- text -}}'
{% endmacro %}
```

---

To implement a unit testing macro corresponding to the `to_literal` macro we can create a macro to test our original macro in `integration_tests/macros/test_to_literal.sql`.

Then we call the `to_literal` macro in the testing macro. and if the result isn’t the same as expected, we raise an error using the [exceptions.raise_compiler_error macro](https://docs.getdbt.com/reference/dbt-jinja-functions/exceptions).

```sql
-- integration_tests/macros/test_to_literal.sql
{% test_to_literal() %}
	{% = dbt_sample_package.to_literal('test string') %}
	{% if result != "'test string'" %}
		{{ exceptions.raise_compiler_error('The test is failed') }}
	{% endif %}
{% endmacro %}
```

---

By doing that, we can call the testing macro in the dbt project of integration tests using `dbt run-operation`.

```sql

> dbt -operation test_to_literal

Reviewer: ```suggestion
dbt run-operation test_to_literal
```

```

---

If we want to run all tests with a single command, it would be good to bundle them in a macro. Moreover, we can call the macro with `dbt run-operation`.

```sql
-- integration_tests/macros/run_unit_tests.sql
{% macro run_unit_tests() %}

> 	{% do test_to_literal() %}

Reviewer: ```suggestion
	{% do test_to_literal() %}
	{% do another_test() %}
```

Or something like this to show that "bundling" power here

Author: ahh excellent idea

```sql
-- integration_tests/macros/test_to_literal.sql
{% macro test_to_literal() %}
	{{ return(adapter.dispatch('test_to_literal', 'integration_tests')(text)) }}
{% endmacro %}

{% macro default__test_to_literal() %}
	{% result = dbt_sample_package.to_literal('test string') %}
	{% if result != "'test string'" %}
		{{ exceptions.raise_compiler_error('The test is failed') }}
	{% endif %}
{% endmacro %}

{% macro postgres__test_to_literal() %}
	{% result = dbt_sample_package.to_literal('test string') %}
	{% if result != "E'test string'" %}
		{{ exceptions.raise_compiler_error('The test is failed') }}
	{% endif %}
{% endmacro%}
```

---

We can then select unit tests based on the specified adapter. Let’s assume we have different dbt profiles corresponding to BigQuery and postgres. By specifying a dbt profile based on the adapter, we can select what testing macros are called internally.

```sql
# Run unit tests on BigQuery
# `default__test_to_literal` is internally called.

> dbt -operation run_unit_tests --profile bigquery

Reviewer: ```suggestion
dbt run-operation run_unit_tests --profile bigquery
```

# Run unit tests on postgres
# `postgres__test_to_literal` is internally called.

> dbt -operation run_unit_tests --profile postgres

Reviewer: ```suggestion
dbt run-operation run_unit_tests --profile postgres
```

---

> title: "An introduction to unit testing your dbt Packages"

Reviewer 2: Just curious. Apparently, there is a spelling inconsistency on "dbt package(s)" or "dbt Package(s)". Which is better for the article?

description: "Traditionally, integration tests have been the primary strategy for testing dbt Packages. In this post, Yu Ishikawa walks us through adding in unit testing as well."
slug: unit-testing-dbt-packages
authors: [yu_ishikawa]
tags: [dbt tutorials]
hide_table_of_contents: false

date: 2022-08-25
is_featured: true
---

> _Editors note - this post assumes working knowledge of dbt Package development. For an introduction to dbt Packages check out [So You Want to Build a dbt Package](https://docs.getdbt.com/blog/so-you-want-to-build-a-package)._

Reviewer: ```suggestion
_Editor's note - This post assumes working knowledge of dbt Package development. For an introduction to dbt Packages check out [So You Want to Build a dbt Package](https://docs.getdbt.com/blog/so-you-want-to-build-a-package)._
```

	{% endif %}
{% endmacro %}
```

---

By doing that, we can call the testing macro in the dbt project of integration tests using `dbt run-operation`.

```sql
<<<<<<< HEAD
dbt run-operation test_to_literal
=======
dbt -operation test_to_literal
>>>>>>> 87760d85 (load unit testing post)
```

---

If we want to run all tests with a single command, it would be good to bundle them in a macro. Moreover, we can call the macro with `dbt run-operation`.

```sql
-- integration_tests/macros/run_unit_tests.sql
{% macro run_unit_tests() %}
	{% do test_to_literal() %}
<<<<<<< HEAD
	{% do another_test() %}
=======
>>>>>>> 87760d85 (load unit testing post)
{% endmacro %}
```

---

## Unit tests for multiple adapters

Your dbt package may support multiple adapters. If you are a postgres user, you understand that the preceding `to_literal` macro doesn’t work on postgres because the expression to deal with string literal is different. So, we have to implement a macro to handle a special case of postgres. Now, we implement the subsequent macro called `postgres__to_literal` in `macros/to_literal.sql` in addition to the implementation above.

```sql
-- macros/to_literal.sql
{% macro to_literal(text) %}

> 	{{ return(adapter.dispatch('literal_string', 'dbt_sample_package')(text)) }}

Reviewer 2: I made a minor mistake in the code block.

```suggestion
	{{ return(adapter.dispatch('to_literal', 'dbt_sample_package')(text)) }}
```
